lib/manager/gradle/index.ts

```diff
@@ -29,7 +29,11 @@ export function updateDependency(
 export const language = LANGUAGE_JAVA;
 
 export const defaultConfig = {
-  fileMatch: ['\\.gradle(\\.kts)?$', '(^|/)gradle.properties$'],
+  fileMatch: [
+    '\\.gradle(\\.kts)?$',
+    '(^|/)gradle.properties$',
+    '\\.versions\\.toml$',
```

Comment:
The file hasn't to include `.versions`.

I think it is not optimal to search for all toml files in the project, we should parse the exact toml files passed on the `settings.gradle.kts` plus the default file (`gradle/libs.versions.kts`).

Reply:
The problem here is that you can even write a valid TOML in a `whate.ver` file, and Gradle would parse it anyway, and you can define them in a zillion ways. So, either we give up the current shallow parser and write a Kotlin library intercepting the Gradle API configuration, or we support the large majority of cases, documenting what we don't support and adding further details upon requests.

I think it is fine to match all TOML files, but wouldn't get past that on the first version. 99.9% of gradle projects will IMHO use the default `gradle/libs.versions.toml`.

Reply:
Deep parser is declined. I think we should only parse default file with file match.

We can later extend the kotlin / Gradle parser, so it finds and extract custom catalog file references.

Users can always override the default file match, if they have non-normal special custom files
.

Reply:
We will deprecate the deep parser soon, so we can remove it in a future major bump.

Reply:
Personally, I am using always two files, the default one and another, so I prefer the approach of searching for all toml files instead of only the default one.

Reply:
But your custom file is somehow referenced in Gradle config? We should try to detect it from there.

Parsing all toml files would have too many false positives for other other project types which also use toml files.

Reply:
Yeah, all custom locations has to be added to Gradle, i.e:

```kotlin
dependencyResolutionManagement {
    versionCatalogs {
        // declares an additional catalog, named 'testLibs', from the 'test-libs.versions.toml' file
        create("testLibs") {
            from(files("gradle/test-libs.versions.toml"))
        }
    }
}
```

Reply:
So we should detect and extract those instead of matching all toml files.

Reply:
Possibilities:
1. Detect any custom file name as long as there's a convention for referencing it from one of the files we *do* match by default, or
2. Scan all tool files relative to build.gradle
3. Support custom file names but only if configured in fileMatch

I think we can start with 3.

I agree that scanning every toml file in every repo "just in case" is too inefficient.

Reply:
Custom locations can be only declared in `settings.gradle.kts` files, so searching in those files and extracting all `.toml` strings inside the block `versionCatalogs` should be enough.

Maybe those strings can be in some variable outside that block, but I think it will be strange.

Reply:
We shuld think of what forms we want to support. Since build specs are Kotlin code, you can easily write something like:
```kotlin
dependencyResolutionManagement {
    repositories {
        mavenCentral()
    }
    val catalogs = listOf("some", "custom", "names")
    versionCatalogs {
        catalogs.forEach {
            create("aCatalogNamed${it.capitalize()}") {
                from(files("$it/$it-libs.toml"))
            }
        }
    }
}
```

Good luck supporting this and hundreds of other cases without resorting to the Gradle API.

In this case, I think that we should adhere to the principle of minimum surprise. Having files hardcoded that work and the same files defined non-hardcoded that don't work would be somewhat surprising to me.

TOML scanning could be slow, maybe (I don't know which performance requirements there are), but I don't expect many false positives: besides being a TOML, you must define stuff in the way Gradle expects. We might scan all TOML, and consider only those that define at least one section among `versions`, `libraries`, and `plugins`. Is there a section name overlap with other toolchains?

Reply:
We won't use the gradle API. We should support parsing the most common conventions to reference other files - enough to cover hopefully all use cases. If there are users who insist on using a syntax or approach we consider too obscure or too difficult to parse, we're ok to "lose" them as a user rather than get into arguments about whether what they're doing is allowed/valid.

Reply:
We consider it a "false positive" situation if we e.g. scan 100 toml files to find every valid gradle versions file, although we haven't quantified exactly how inefficient that really is. I think right now of you ask "what percentage of .toml files on GitHub are gradle's new version catalog?" then it would be pretty low?

Reply:
@rarkins 
Maybe we should scan for gradle TOML files only if a `^build.gradle(\.kts)?$` file is found -- if that could be implemented easily.
Another reasonable strategy is match only TOML files in the `gradle` folder, with a `gradle/*.toml` pattern. It still makes more sense to me than attempting to scan the configuration.

We could also roll out a simple solution initially and see what the community asks. As a first step, I think we must match the default catalog, everything else would be a nice plus. It's better to begin with support for a subset of the features and small extensions, rather than trying to support everything from the start and being at risk of regressions, IMHO.

Reply:
In the end, my suggestion (and I updated the code to reflect it is to match):
1. all TOML files inside the `gradle` folder.
2. all `.versions.toml` files.

I believe this solution reaches a nice spot between low false-positive count and configuration options.

Reply:
@viceice @rarkins if you agree that the current setup is reasonable and easy to extend in the future, please mark the conversation as resolved.